### DISCUSSION TITLE
Tabulated data; inclussion and usage

### DISCUSSION THREAD

**Discussion by Irvise** [thelfer/tfel#89] [Q&A] [answered]
Hi, I am getting started with Tfel/Mfront

I already tried to post this discussion, but it seems it did not pass the Github filter...

My question is regarding tabulated data. I am doing a fairly complex analysis (more on that later) and I would like to be able to include the same data I have available. The data for the properties comes in tabulated form ala:

Young | Temp
--- | ---
100 | 20
50 | 200
... | ...

How does one work with this type of data? I have not seen any documentation about it anywhere (though I must be honest, I have yet a lot to read). Do I have to create an algebraic representation of the data for it to be processed by Mfront?

The analysis I am trying to do is a elasto-visco-plastic (maybe damage too) in Code_Aster and Abaqus. I am using Abaqus as the primary program and all of its built in tools. The only addition I had to make to Abaqus is the inclusion of a customised CREEP subroutine describing the power law; with dependence on temperature and stress. Code_Aster does not provide much flexibility when it comes to creep nor combined isotropic and kinematic hardening (though VISC_CIN2_CHAB and HAYHURST do come close).

Thank you for your time,
Fernando Oleo Blanco

P.S.: I will open another discussion talking about my case in more detail, this is just for the tabulated data.

**Comment by thelfer** [comment, marked as answer]
To be honest, material properties are poorly documented in english (The french documentation is a bit better).

For your problem, there is built-in feature to handle material properties given by tabulated data. You may want to open an issue about that.

The main reason is that one must handle how data are interpolated and extrapolated. `TFEL/Math` provides cubic spline and kriging for example. But no support for linear interpolation !

You'll find attached two examples of how to implement that.  The first one is compatible with `TFEL-4.0`. The second one shall compile with previous versions:
  
- [TestYoungModulus.zip](https://github.com/thelfer/tfel/files/7908265/TestYoungModulus.zip)
- [TestYoungModulus-without-qt.zip](https://github.com/thelfer/tfel/files/7908266/TestYoungModulus-without-qt.zip)

**Reply by Irvise** [reply]
Oh great!

First things first, thank you for the quick reply :)

Second, great that math does have a way to work with tabulated data. I will give it a look.

However, as you point out... No linear interpolation? That is a bit strange, specially when you have splines and kriging x) I would like to use linear interpolation mainly to mimic the behaviour that Abaqus has. This would help me validate the model. To me it would also be important to have he data "expanded" from its original range. CA allows constant interpolation or linear interpolation outside of the range. Abaqus has constant interpolation outside of the range.

I will read the french sources. I am not a french speaker, but I can manage :P 

Regards,
Fer

Edit: tabulation->interpolation

**Reply by thelfer** [reply]
> However, as you point out... No linear tabulation? That is a bit strange, specially when you have splines and kriging x) I would like to use linear interpolation mainly to mimic the behaviour that Abaqus has.

Yes. I probably though that was too easy to implement, as in the two examples....

The other thing is that I do believe that a dedicated support for tabulated data shall be implemented...

**Reply by Irvise** [reply]
I do not know how Tfel works inside. But since it is using newer C++ features, it may be able to define a "function" that is just a tabular datatype that can digest data and output a "function" type. The way it works would just be internal state (linear, quadratic, etc)...

My two cents :P

**Reply by thelfer** [reply]
Yes of course. But the question is about efficiency.
For example, we could precompute coefficients of the cubic spline at compile time, etc...

**Reply by Irvise** [reply]
Oh, I did not think about that. A cubic may indeed may be more efficient than a linear one... I suppose a linear interpolation has O(n) or O(log n) complexity, compared to a precomputed O(1) for cubic...

**Comment by thelfer** [comment]
> The analysis I am trying to do is a elasto-visco-plastic (maybe damage too) in Code_Aster and Abaqus. I am using Abaqus as the primary program and all of its built in tools. The only addition I had to make to Abaqus is the inclusion of a customised CREEP subroutine describing the power law; with dependence on temperature and stress. Code_Aster does not provide much flexibility when it comes to creep nor combined isotropic and kinematic hardening (though VISC_CIN2_CHAB and HAYHURST do come close).

Di you have a look at the [`StandardElastoViscoPlasticityBrick`](https://thelfer.github.io/tfel/web/StandardElastoViscoPlasticityBrick.html) ?

**Reply by Irvise** [reply]
Yes I have.

In terms of elasto-plasticity it seems to cover my case... more or less. The "issue" is that I would like to replicate the exact behaviour that Abaqus has internally. For plasticity I have a [combined parameter hardening](https://abaqus-docs.mit.edu/2017/English/SIMACAEKEYRefMap/simakey-r-plastic.htm#simakey-r-plastic__simakey-r-plastic-s-datadesc3) entry (already given by the data I was handed). It is also temperature dependent, so I do not know if I can use the brick...

In terms of viscosity/creep, the block does not provide what I need, so that would have to be done elsewhere.

I suppose I have to implement everything under the Implicit DSL. I was thinking on taking some public .mfront file and using it as a base, I think the Hayhurst is open...

Regards,
Fer

**Reply by thelfer** [reply]
> In terms of elasto-plasticity it seems to cover my case... more or less. The "issue" is that I would like to replicate the exact behaviour that Abaqus has internally. For plasticity I have a [combined parameter hardening](https://abaqus-docs.mit.edu/2017/English/SIMACAEKEYRefMap/simakey-r-plastic.htm#simakey-r-plastic__simakey-r-plastic-s-datadesc3) entry (already given by the data I was handed). It is also temperature dependent, so I do not know if I can use the brick...

Well, I think you can :)

**Reply by thelfer** [reply]
> In terms of viscosity/creep, the block does not provide what I need, so that would have to be done elsewhere

Tell me what you need. Maybe we could add it to the brick.

**Reply by thelfer** [reply]
> I suppose I have to implement everything under the Implicit DSL. I was thinking on taking some public .mfront file and using it as a base, I think the Hayhurst is open...

Yes it is. It may a bit outdated though.

You may also find interesting material on [this draft](https://github.com/thelfer/tfel-doc/tree/master/TrainingCourse/Tutorial). It still needs a lot of work. Please find attached a pdf version.

[Tutorial.zip](https://github.com/thelfer/tfel/files/7908580/Tutorial.zip)

**Reply by Irvise** [reply]
This is a fairly long term project for me. In reality, I just have to work with Abaqus, and that is going more or less fine. The CA case is for myself and to use more powerful hardware. I expected CA to be able to handle the materials, but it does not seem like it does, so I am falling back to Mfront :) Since this is "just" for me, I will take it more slowly, but I will keep on reporting and commenting/helping :)

**Reply by Irvise** [reply]
If you are interested, here is some reading :P 

Some [copper creep](https://inis.iaea.org/collection/NCLCollectionStore/_Public/43/084/43084402.pdf). Mine is Cu-HCOF, basically pure copper, without P.
And [some creep models](https://sundoc.bibliothek.uni-halle.de/habil-online/06/06H055/t3.pdf)

**Reply by thelfer** [reply]
No problem, take your time. Also, MFront does work fine with Abaqus. Maybe you'll find it more convenient to work with MFront rather than raw UMAT :)

**Reply by Irvise** [reply]
That is exactly why I am looking into MFront. UMAT may be very flexible, but it is all done by hand... The CREEP subroutine is already enough for me :) Also, MFront works for both programs I am interested in, which is a big plus. I do not know how far UMAT compatibility goes in CA.

**Reply by thelfer** [reply]
> If you are interested, here is some reading :P
> 
> Some [copper creep](https://inis.iaea.org/collection/NCLCollectionStore/_Public/43/084/43084402.pdf). Mine is Cu-HCOF, basically pure copper, without P. And [some creep models](https://sundoc.bibliothek.uni-halle.de/habil-online/06/06H055/t3.pdf)

Well, I just had a quick look. Seems quite standard, doesn't it ?
Let me know which models are the more suitable for you (which equation in those documents)

**Reply by thelfer** [reply]
> I do not know how far UMAT compatibility goes in CA.

To be honest, I don't know either !

**Reply by Irvise** [reply]
From what I have read, hyperbolic expressions are good for fracture dynamics, since they behave exponentially with stresses. I have not read much more regarding the creep/viscosity terms used in CA regarding the Lematrie model, nor the Norton, [Chaboche](https://www.code-aster.org/V2/doc/default/en/man_r/r5/r5.03.04.pdf) and [Taheri](https://www.code-aster.org/V2/doc/default/en/man_r/r5/r5.03.05.pdf). But they seem to be simple (though useful) approximations to viscous flow.

Most literature that I read about creep in copper seems to be in favour of the power law model (I will link a reference later today or tomorrow). Also, the hyperbolicsine creep in CA does not have a temperature dependence term (Arrhenius), while Abaqus does have. MFront also does not have the Arrhenius term in the Elastoviscoplastic block.

Regards,
Fer

P.S.: I know I am a bit pedantic regarding temperature dependence, but it is an integral part of the study I am doing.

**Reply by thelfer** [reply]
> Most literature that I read about creep in copper seems to be in favour of the power law model (I will link a reference later today or tomorrow). Also, the hyperbolicsine creep in CA does not have a temperature dependence term (Arrhenius), while Abaqus does have. MFront also does not have the Arrhenius term in the Elastoviscoplastic block.

The documentation of the brick is probably not good enough. There is absolutely not difficulty to introduce the temperature dependance.
Have a look at page 10 in the tutorial that I sent you. You can also use external MFront files to implement the material coefficients.

**Reply by Irvise** [reply]
Sorry that I have not answered for a while here. Just had a discussion with my supervisor. We have settled on the Norton-Bailey formulation. The paper with the data that we are going to use is [this one](https://experts.illinois.edu/en/publications/modeling-creep-and-fatigue-of-copper-alloys). The open access version is [available here](http://ccc.illinois.edu/PDF%20Files/Reports/Cu_All_CCC_mar_21_00.pdf). We will be working with Cu-OFHC and CuCrZr (first entry on the table); and we will be using their models. Better creep analysis are currently being carried out in the UK, but they will take too much time for what we need to get done.

While we would like to model primary and secondary (most often than not also called minimum rate) creep/creep rate. The Norton-Bailey is mostly designed for the secondary creep (with the parameters tweaked to fit primary, specifically the time exponent). There is a paper done for the Swedish nuclear body that does model primary creep in Cu-OFP quite nicely, but it uses a pure physical description of the crystalline dynamics, which is too much for us.

I am leaving this here for reference, not that I want you to implement it. I also need to give a read to the brick that you linked. I have been busy implementing the CREEP subroutine.

Regards,
Fer

P.S; I noticed that the singularity container for CA has Mfront but version 3.4. I will have to compile the newer version there :) And I still haven't gotten around trying what we talked about in MSYS2, sorry!

**Reply by thelfer** [reply]
You want to implement this expression ?

ε (s-1) = 1.43×1010 exp (-197,000/8.31 T(K)) (σ(MPa))2.5 (t(s))-0.9

**Reply by Irvise** [reply]
For the CuCrZr we decided to add up the primary and secondary creep expressions. While that is not correct, the primary creep rate is a couple orders of magnitude greater than the second and it decays with time, so it would be something like:

ε (s-1) = (exp (-197,000/8.31 * T(K)) * (7.56x10^11*(σ(MPa))^3*(t(s))^(-0.92) + XXX)

Where XXX is an interpolation of the second line entry and the third line entry for CuCrZr (their difference is the temperature of the tests). Without the Arrhenius term.

For Cu-OFHC we are just going to use the secondary creep expression listed there (first material entry in the table). Maybe we already see a lot of stress relaxation, which will require the correct data that will be generated in the UK, otherwise, we will have to wait and see.

Edit: we know that the footnote says that we should change from primary to secondary when their strains are the same. However, for stability's sake, we do not want to include an if.

**Reply by thelfer** [reply]
For the fun, I just implemented the primary creep. This requires a bit of reworking the initial expression as time hardening relation is not a constitutive equation. This is described in the attached note. Adding the secondary creep is trivial.

We could even discuss how to make things even better by replacing material properties with parameters, computing the elastic properties in the behaviour, etc..

Pretty sure that in the end, the implementation is nicer in `MFront` than in the `CREEP` module of `Abaqus`.

[TimeHardening.zip](https://github.com/thelfer/tfel/files/7948475/TimeHardening.zip)
.

**Reply by Irvise** [reply]
Hi @thelfer,

I took a look at the document you included. Just a minor comment, the first equation, as I have seen it mostly mentioned in the literature is for the creep rate, not the creep. So the creep rate has a time hardening constant (in your document `n_t`) without `-1`. But this is not really important, since that would just require shuffling some coefficients around or making small corrections.
Otherwise, thank you for adding the `.mfront` file with the documented explanation for the derivation, it was nice to see how you approached the problem.

The reason I have not responded earlier is because I started to do some sanity checks on the data I included in an earlier message (the big table one). And after digging more and more on the references and literature, and getting results that were most likely wrong, the strategy has changed.

We have switched to a parametric analysis... It is not great, but it is the best we can do now. So my time has been spent setting up the parametric analysis and data ingestion system... We hope that our results will be an aid to a research team in UK and Italy, who are the ones producing the actual data.

On a personal level, I still hope to develop the material definitions in `MFront` and work with it in `Code_Aster`; since `CA` really needs some extra work on that front, but that has been pushed further in time do to the new load of work.

I am sorry for the delay, but thank you very much none the less! Best regards,
Fer

**Reply by thelfer** [reply]
@Irvise Thanks for the feed-backs. Let me know if you ever need more help !

**Comment by thelfer** [comment]
BTW, do not hesitate to use `mfront` documentation. For example:

```bash
$ mfront --help-keyword=MaterialLaw:@Law
$ mfront --help-keyword=MaterialLaw:@Function
```
